### PR TITLE
SDK enhancements

### LINKED ISSUE #15: Video Upload error
It is showing statusNumber =1500 ,statusCode =SERVER_ERROR and message= Signature genertion Failed

### PR DESCRIPTION
**Pull request by AniVerma17**
- Add following features in the SDK:
    - Upload policy to set constraints on upload requests for device conditions, retry mechanism etc.
    - URL constructor methods for adaptive video streaming, adding raw transform options as a string, and image loading in the responsive manner according to the target view.
    - Upload preprocessors to transform images and video before uploading.
    - Library modules for providing extensions for 3rd party image loading libraries.
- Migrate the uploader methods and API calls to the V2 upload API.
- Remove the deprecated overlay transformation options from URL constructor.

Also resolves #15 and #13 .

### CODE REVIEW

File: imagekit/src/test/java/com/imagekit/android/UnitTests.kt [hunk 1]
```diff
@@ -518,776 +519,7 @@ class UnitTests {
...
-            .overlayText(overlayText = "overlay made easy")
-            .overlayTextTypography(overlayTextTypography = OverlayTextTypography.BOLD_ITALICS)
-            .create()
-        assertEquals(
-            actual,
-            "https://ik.imagekit.io/demo/tr:ot-overlay+made+easy,ott-bi/medium_cafe_B1iTdD0C.jpg?${ImageKit.IK_VERSION_KEY}=android-${BuildConfig.API_VERSION}"
-        )
-    }
-
-    @Test
```
Comment (imagekitio) [inline imagekit/src/test/java/com/imagekit/android/UnitTests.kt:1269]: We need to add test cases for `raw`.

**Review by imagekitio — CHANGES_REQUESTED** [review]
Functionality wise this PR looks good, but we need to add more test cases for new features and modify assertions in upload test cases.
## Upload test cases
We are mocking response and asserting SDK output. This is not enough as this is only testing that SDK is correctly parsing response. In addition to this, we need to assert the okHTTP request body somehow to ensure SDK is issuing correct request. We are doing same in web client side SDK - https://github.com/imagekit-developer/imagekit-javascript/blob/master/test/upload.js

## New feature test cases
Let's discuss on call and decide what to assert in all of them:
* Upload policy
* Image/video pre processing step
* Responsive images logic
* Third party integrations i.e. glide, coil etc

File: imagekit/src/test/java/com/imagekit/android/URLGenerationTests.kt [hunk 2]
```diff
@@ -222,7 +221,7 @@ class URLGenerationTests {
     @Test
     fun urlConstructionCustomQueryParameterWithTransforms() {
         val expectedTransformation =
-            "https://ik.imagekit.io/demo/medium_cafe_B1iTdD0C.jpg?${ImageKit.IK_VERSION_KEY}=android-${BuildConfig.API_VERSION}&x-test-header=Test&tr=h-300,w-300:rt-90"
+            "https://ik.imagekit.io/demo/medium_cafe_B1iTdD0C.jpg?x-test-header=Test&tr=h-300,w-300:rt-90"
         val actualTransformation =
```
Comment (imagekitio) [inline imagekit/src/test/java/com/imagekit/android/URLGenerationTests.kt:226]: Need to have test cases for adaptive bitrate URL generation and in general all new video related parameters that are added now.

**Review by imagekitio — CHANGES_REQUESTED** [review]

File: imagekit/src/test/java/com/imagekit/android/UnitTests.kt (hunk 1 above)
Comment (AniVerma17) [reply imagekit/src/test/java/com/imagekit/android/UnitTests.kt:1269]: Added in #33

File: imagekit/src/test/java/com/imagekit/android/URLGenerationTests.kt (hunk 2 above)
Comment (AniVerma17) [reply imagekit/src/test/java/com/imagekit/android/URLGenerationTests.kt:226]: Implemented in #33

**Comment by AniVerma17** [comment]
> Functionality wise this PR looks good, but we need to add more test cases for new features and modify assertions in upload test cases.
> 
> ## Upload test cases
> We are mocking response and asserting SDK output. This is not enough as this is only testing that SDK is correctly parsing response. In addition to this, we need to assert the okHTTP request body somehow to ensure SDK is issuing correct request. We are doing same in web client side SDK - https://github.com/imagekit-developer/imagekit-javascript/blob/master/test/upload.js
> 
> ## New feature test cases
> Let's discuss on call and decide what to assert in all of them:
> 
> * Upload policy
> * Image/video pre processing step
> * Responsive images logic
> * Third party integrations i.e. glide, coil etc

The unit tests for these features have been added in PR #33. Following assertions are covered in these tests:

- Upload API calls: Assertions for individual parameters in the multipart request body generated in the SDK, captured in a MockWebServer instance.
- Upload policy: device constraints to allow or disallow an upload request, using Roboelectric to mock real device conditions
- Upload preprocessors: Image preprocessors (max dimensions limiter, crop and rotate) and output format check
- Responsive images logic: assertion of dimension and parameters in the URL to match those of the mock view passed to it
- Third party integrations: Tests for Coil and Fresco extensions to validate the URL in their request objects to match the SDK output

**Pull request merged**